Source/common/SNTXPCUnprivilegedControlInterface.h

```diff
@@ -51,6 +51,11 @@ struct RuleCounts {
 - (void)databaseEventCount:(void (^)(int64_t count))reply;
 - (void)staticRuleCount:(void (^)(int64_t count))reply;
 
+///
+/// Standalone rules
+///
+- (void)addStandaloneRule:(SNTRule *)rule reply:(void (^)(NSError *))reply;
```

Comment:
I don’t think we want to expose adding arbitrary rules over the unprivileged interface. Instead, we could have the daemon drive the flow, sending a message to the GUI for auth.

Reply:
Admittedly I went back and forth on this. What were you thinking the flow would look like?

The main reason I added this was to keep the flow simpler then the santad flows I sketched out. Felt like I had to keep state in the daemon which felt wrong. 

If you enable standalone mode then our flow looks like this.

```mermaid
sequenceDiagram
santad ->> gui: We blocked a binary pop up message and display approve button
gui ->> gui: User clicks approve
gui ->> gui: Get auth from user
gui -->> santad: addStandaloneRule
```

If the user clicks ignore then nothing happens.

Reply:
> The main reason I added this was to keep the flow simpler then the santad flows I sketched out. Felt like I had to keep state in the daemon which felt wrong.

Having the GUI push in rules is simpler, but I don't think we want to expose adding rules from an unprivileged interface. With that as a requirement, we will need to store state with the daemon.

sequenceDiagram
santad ->> santad: blocks an executable - checks if standalone mode and desired auth is available
santad ->> rules.db: adds a pending rule
santad ->> gui: prompt user and display approve button
gui ->> gui: User clicks approve
gui ->> gui: Get auth from user
gui ->> santad: activateStandaloneRule:
santad ->> rules.db: converts pending rule to active

The idea of a "pending rule" is new and we would need to think though the implications.

The other thing we need to consider is the key used with activateStandaloneRule. The obvious choice is to just use the blockable event. This is probably fine. However, it does not strongly guarantee that a user approved the event. We could consider using a hard to guess nonce, or figure out if LocalAuthentication can mint an attestation that we verify in the activateStandaloneRule: implementation.

Reply:
Chatted offline - using a reply block would eliminate the need for exposing an additional interface in the daemon. This is a great idea!